### PR TITLE
Congestion control

### PR DESCRIPTION
**Pull request by wykurz**

### CODE REVIEW

File: congestion/src/measurement.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::testing::CollectingSink;
+
+    // tests that touch the global sink must run serially within the module;
+    // each test clears the sink on exit to keep state isolated.
+
```
[the remark this answers is not in the document]
Comment (wykurz) [reply congestion/src/measurement.rs]: Fixed in b1618bf → bfa8dbd: introduced a shared `tokio::sync::Mutex` (`measurement::SINK_GUARD`) exposed at the crate boundary and guarded every test that mutates the process-wide sink with it. `measurement::tests` uses `blocking_lock()` (sync `#[test]`), `control_loop::tests` uses `.lock().await` (async `#[tokio::test]`), and both share the same static so cross-module races are prevented too.

File: congestion/src/control_loop.rs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    async fn routing_sink_integrates_with_global_probe_api() {
+        let mut builder = RoutingSinkBuilder::new();
+        let mut meta_rx = builder.metadata_receiver();
+        let sink = builder.build();
+        install_sample_sink(std::sync::Arc::new(sink));
+        Probe::start_metadata().complete_ok(0);
+        let s = meta_rx.recv().await.expect("sample flowed through");
+        assert_eq!(s.bytes, 0);
+        clear_sample_sink();
+    }
```
[the remark this answers is not in the document]
Comment (wykurz) [reply congestion/src/control_loop.rs:455]: Fixed in bfa8dbd: this test now takes `crate::measurement::SINK_GUARD.lock().await` before touching the process-wide sink. Same shared mutex as `measurement::tests` so the two modules are mutually serialized.

File: throttle/src/semaphore.rs [hunk 3]
```diff
@@ -1,26 +1,74 @@
...
+    pub fn set_max(&self, value: usize) {
+        let current = self.limit.swap(value, Ordering::AcqRel);
+        self.flag.store(value > 0, Ordering::Release);
+        match value.cmp(&current) {
+            std::cmp::Ordering::Greater => {
+                self.sem.add_permits(value - current);
+            }
+            std::cmp::Ordering::Less => {
+                self.sem.forget_permits(current - value);
+            }
```
[the remark this answers is not in the document]
Comment (wykurz) [reply throttle/src/semaphore.rs:144]: Real limitation, but not one we hit in practice — the adaptive controller is configured with `min_cwnd >= 1`, so the controller never emits a zero-cap decision. Rather than adding a cancellation mechanism that would have no caller today, I tightened the docstring on `Semaphore::set_max` and the public `throttle::set_max_ops_in_flight` in bfa8dbd to explicitly call out that disable applies only to new acquires — waiters already parked inside the semaphore stay parked until permits become available. Callers that need cancellable shutdown should use a higher-level shutdown signal. Happy to revisit if a data-path controller (future Phase 5) needs the zero-transition to be tight.

File: congestion/src/measurement.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    *SINK.write().expect("sample sink lock poisoned") = Some(sink);
+}
+
+/// Remove the current sink, if any. After this call, probes are no-ops again.
+pub fn clear_sample_sink() {
+    *SINK.write().expect("sample sink lock poisoned") = None;
+}
+
+fn emit(kind: ResourceKind, sample: &Sample) {
+    if let Some(sink) = SINK.read().expect("sample sink lock poisoned").as_ref() {
```
[the remark this answers is not in the document]
Comment (wykurz) [reply congestion/src/measurement.rs]: Fixed locally (pending push): `emit` now clones the `Arc<dyn SampleSink>` out of the RwLock before dispatching, so a slow sink no longer blocks `install`/`clear_sample_sink` and a sink that re-enters the sink API from within `record` cannot self-deadlock.

File: throttle/src/semaphore.rs [hunk 5]
```diff
@@ -1,26 +1,82 @@
...
 
     pub fn setup(&self, value: usize) {
         self.flag.store(value > 0, Ordering::Release);
+        self.sem.forget_permits(self.sem.available_permits());
+        self.limit.store(value, Ordering::Release);
         if value == 0 {
             return;
         }
-        self.sem.forget_permits(self.sem.available_permits());
         self.sem.add_permits(value);
```
[the remark this answers is not in the document]
Comment (wykurz) [reply throttle/src/semaphore.rs]: Fixed locally (pending push): `setup` now flips `flag` to `false` first, reconfigures permits, then flips back to `true` only after `add_permits` has run. The permit-free window between `forget_permits` and `add_permits` can no longer be observed by a concurrent acquirer as "enabled but empty."

File: throttle/src/semaphore.rs [hunk 6]
```diff
@@ -1,26 +1,82 @@
...
+    /// higher-level shutdown signal; the adaptive controller does not rely
+    /// on zero transitions in practice (its minimum cwnd is configured
+    /// `>= 1`).
+    pub fn set_max(&self, value: usize) {
+        let current = self.limit.swap(value, Ordering::AcqRel);
+        self.flag.store(value > 0, Ordering::Release);
+        match value.cmp(&current) {
+            std::cmp::Ordering::Greater => {
+                self.sem.add_permits(value - current);
+            }
```
[the remark this answers is not in the document]
Comment (wykurz) [reply throttle/src/semaphore.rs:141]: Fixed locally (pending push): `set_max` now orders updates so that (a) for a disable transition, `flag` is flipped to `false` *before* `forget_permits` runs, and (b) for any enable/grow transition, permits are added/forgotten *before* `flag` is flipped to `true`. The "flag=true with zero permits" window is eliminated for 0 → N transitions.

File: common/src/rm.rs [hunk 7]
```diff
@@ -269,19 +269,34 @@ async fn rm_internal(
...
+            .map_err(|err| Error::new(err, Default::default()))?;
+        let Some(entry) = maybe_entry else {
+            probe.discard();
+            drop(ops_permit);
+            break;
+        };
         let entry_path = entry.path();
         // check entry type for filter matching and skip counting
         let entry_file_type = entry.file_type().await.ok();
+        probe.complete_ok(0);
```
[the remark this answers is not in the document]
Comment (wykurz) [reply common/src/rm.rs]: Fixed locally (pending push): `rm_internal` now matches on `file_type().await` and calls `probe.discard()` on Err rather than `probe.complete_ok(0)`. Discard (vs `complete(Outcome::Error)`) chosen because Vegas ignores Outcome today and error-path latency is not a useful congestion signal anyway — cleanest to keep it out of the sample stream entirely.

File: common/src/copy.rs [hunk 8]
```diff
@@ -745,21 +745,33 @@ async fn copy_internal(
...
+            probe.discard();
+            drop(ops_permit);
+            break;
+        };
         let entry_path = entry.path();
         let entry_name = entry_path.file_name().unwrap();
         let dst_path = dst.join(entry_name);
         // check entry type for filter matching and skip counting
         let entry_file_type = entry.file_type().await.ok();
+        probe.complete_ok(0);
```
[the remark this answers is not in the document]
Comment (wykurz) [reply common/src/copy.rs]: Fixed locally (pending push): same fix as rm_internal — `probe.discard()` on `file_type()` error so error-path samples don't feed the controller.

File: common/src/link.rs [hunk 9]
```diff
@@ -512,21 +512,33 @@ async fn link_internal(
...
+            probe.discard();
+            drop(ops_permit);
+            break;
+        };
         let cwd_path = cwd.to_owned();
         let entry_path = src_entry.path();
         let entry_name = entry_path.file_name().unwrap();
         // check entry type for filter matching and dry-run reporting
         let entry_file_type = src_entry.file_type().await.ok();
+        probe.complete_ok(0);
```
[the remark this answers is not in the document]
Comment (wykurz) [reply common/src/link.rs]: Fixed locally (pending push): same fix as rm_internal/copy_internal — `probe.discard()` on `file_type()` error. Went with discard over `Outcome::Error` since Vegas ignores Outcome today; either would be correct.

File: throttle/src/semaphore.rs [hunk 10]
```diff
@@ -55,3 +137,145 @@ impl Semaphore {
...
+        let _g3 = sem.acquire().await.unwrap();
+        assert_eq!(sem.sem.available_permits(), 2);
+        // shrink past available: forget_permits only takes what's there.
+        // Held permits are NOT revoked — they return on drop.
+        sem.set_max(1);
+        assert_eq!(sem.sem.available_permits(), 0);
+        // dropping the held permits returns them (over the new cap, by
+        // design — the docstring notes this is intentional).
+        drop(_g1);
+        assert_eq!(sem.sem.available_permits(), 1);
```
[the remark this answers is not in the document]
Comment (wykurz) [reply throttle/src/semaphore.rs]: Fixed locally (pending push): test now drops all three held permits and asserts the final `available_permits == 3` (the documented temporary overshoot past the new cap), rather than the intermediate count after dropping just one. Matches the stated semantics and no longer depends on tokio's internal forget-debt accounting.

File: congestion/src/vegas.rs [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+    fn on_sample(&mut self, sample: &Sample) {
+        // u64 nanos fit any realistic latency; saturate defensively.
+        let latency_ns = u64::try_from(sample.latency().as_nanos()).unwrap_or(u64::MAX);
+        // only advance `observed_at` when a strictly lower value is seen
+        // so the age check reflects when this particular minimum was last
+        // confirmed, not the last time any sample arrived.
+        self.min_latency = Some(match self.min_latency {
+            Some((current, at)) if current <= latency_ns => (current, at),
+            _ => (latency_ns, sample.completed_at),
+        });
```
[the remark this answers is not in the document]
Comment (wykurz) [reply congestion/src/vegas.rs]: Fixed in 74df82f094: updated `VegasController::on_sample` to refresh `observed_at` on equal-latency samples (not just strictly-lower ones). The match arm now returns `(current, at)` only when `latency_ns > current`; at-or-below cases fall through to a fresh `(latency_ns.min(current), sample.completed_at)` tuple. Regression test `equal_min_latency_samples_refresh_observed_at` added.

File: congestion/src/vegas.rs [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+            Some(prev) => self.config.ewma_alpha * mean_ns + (1.0 - self.config.ewma_alpha) * prev,
+            None => mean_ns,
+        });
+        self.tick_sum_latency_ns = 0;
+        self.tick_sample_count = 0;
+        if !is_first_sample_tick
+            && let (Some(ewma), Some((min, _))) = (self.ewma_latency_ns, self.min_latency)
+        {
+            let ratio = ewma / (min as f64);
+            if ratio < self.config.alpha {
```
[the remark this answers is not in the document]
Comment (wykurz) [reply congestion/src/vegas.rs:177]: Fixed in 74df82f094: `on_sample` now clamps `latency_ns` to `>= 1` at ingestion (`u64::try_from(..).unwrap_or(u64::MAX).max(1)`). This keeps `min_latency` out of 0, so the `ewma / min` ratio stays finite under clock-resolution edge cases. Regression test `zero_latency_samples_do_not_collapse_cwnd` added — also seen during the property-test work where `Probe::start+complete_ok` in a tight loop produced sub-nanosecond latencies.

File: common/src/config.rs [hunk 13]
```diff
@@ -166,3 +217,75 @@ impl Default for TracingConfig {
...
+    }
+
+    #[test]
+    fn beta_at_or_below_one_is_rejected() {
+        let mut auto = valid_auto_meta();
+        // keep alpha < beta so the alpha-vs-beta check isn't what trips it
+        auto.alpha = 0.9;
+        auto.beta = 1.0;
+        let err = config_with(auto).validate().unwrap_err();
+        assert!(err.contains("alpha") || err.contains("beta"), "got: {err}");
```
[the remark this answers is not in the document]
Comment (wykurz) [reply common/src/config.rs]: Fixed in 74df82f094: test now uses `alpha=1.05` (passes the `alpha > 1.0` check) with `beta=1.0` so the beta branch is specifically what trips. Assertion tightened to `err.contains("beta")`.

File: congestion/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+//! # Built-in controllers
+//!
+//! - [`NoopController`] — never limits. Default when congestion control is
+//!   disabled.
+//! - [`FixedController`] — honors a static concurrency/rate budget. Mirrors
+//!   the existing manual `--ops-throttle` / `--iops-throttle` knobs and is
+//!   the regression baseline for adaptive algorithms.
+//!
+//! Adaptive controllers (Vegas-style, BBR-style) are planned on this same
+//! trait in later phases.
```
[the remark this answers is not in the document]
Comment (wykurz) [reply congestion/src/lib.rs]: Fixed in 74df82f094: crate docs now list `VegasController` as a shipping adaptive controller (with a brief description of the algorithm) and only BBR-style variants are referenced as future work.

File: throttle/src/lib.rs [hunk 15]
```diff
@@ -214,3 +245,54 @@ pub async fn run_iops_replenish_thread(replenish: usize, interval: std::time::Du
...
+/// Takes effect on the next iteration of the replenish loop started by
+/// [`run_ops_replenish_thread`] — no loop restart, no permits forcibly
+/// drained. Setting `value = 0` pauses replenishment (tokens already in
+/// the bucket will be consumed but no new ones will be added).
+///
+/// Intended for congestion-control layers that translate a Controller's
+/// decisions into dynamic rate targets. The ops-throttle must already
+/// have been initialized via [`init_ops_tokens`] with a non-zero value;
+/// otherwise the replenish loop will have exited and this call is a no-op.
+pub fn set_ops_replenish(value: usize) {
```
[the remark this answers is not in the document]
Comment (wykurz) [reply throttle/src/lib.rs:308]: Fixed in 0ad217f1f4: updated the docs to list the two actual preconditions — a replenish task must be running, and the gate must be enabled (via `init_ops_tokens` with non-zero or `enable_ops_throttle` after a prior disable). The old wording predated removing the flag-check-and-return in `run_replenish_thread`.

File: rcp/tests/remote_tests.rs [hunk 16]
```diff
@@ -4698,3 +4698,71 @@ fn test_remote_skip_specials() {
     assert!(dst_path.join("file.txt").exists());
     assert!(!dst_path.join("test.sock").exists());
 }
+
+/// Verifies that `--auto-meta-throttle` and its tuning flags propagate
+/// from the rcp master to the rcpd processes over the remote protocol.
+///
+/// Before launching each rcpd, the master stderr the full ssh command line
+/// at `Will run remotely: ... /rcpd --role <source|destination> ...`.
```
[the remark this answers is not in the document]
Comment (wykurz) [reply rcp/tests/remote_tests.rs]: Fixed in 0ad217f1f4: `the master stderr the full ssh command line` -> `the master logs the full ssh command line`.
